penetrating shortwave fraction as a 3D-field

mjlosch opened the pull request:
## What changes does this PR introduce?
New feature to allow feedback from bgc to physical fields via shortwave penetration; save some computations for constant (in time) shortwave penetration fraction (current default)

## What is the current behaviour? 
shortwave penetration is been computed locally "on the fly" by `S/R SWFRAC` which does not know anything about the grid coordinates (or indices). As a consequence, it is not possible to locally include any water property (e.g. local turbidity) in the computation that may affect the shortwave penetration. E.g. the Jerlov water type is hard coded to 2 for all applications. Some unfortunate hacks follow in the code, e.g. in `pkg/seaice`, the "swfrac" is computed once in the beginning and put into a common block to lead to potential naming conflicts and a bug (for pressure coordinates), resolved here #744.

## What is the new behaviour 
- introduce 3D field `SWFrac3D` that is precomputed from the grid parameters (`rF`) in `ini_forcing.F`
- this variable (while now initialised within `S/R initialise_fixed`) can be updated each timestep based on water properties calculated by e.g. `pkg/bling` or [Darwin](https://darwinproject.mit.edu); this code in not part of this PR, but there's a suggestion where to insert this.
- this makes the local `pkg/seaice` variable `SEAICE_SWFrac` unnecessary (removed)
- for `pkg/kpp` a clean change is not possible, because the shortwave penetration is computed locally for the dynamic variable `hbl`; an approximation to this is suggested, but that changes results, so a CPP-flag `KPP_USE_SWFRAC3D` needs to be defined to use the new functionality. I suggest to use AD-`1D_ocean_ice_column` to test this, as it also shows that no recomputations are introduced by this and we can actually save a few TAF store directives (results change).
- changes to `pkg/layers/layers_thermodynamics.F` are suggested to be tested in `cfc_example` (no change of results by defining `SHORTWAVE_HEATING`)

## Does this PR introduce a breaking change? 
No

## Other information:
- The KPP code changes are based on code provided by Svetlana Losa. This code was used in  [Pefanis et al. (2020)](https://doi.org/10.1029/2020GL088795)
- This is a Draft PR because I am sure that there will be suggestions to do this better (or maybe not at all?). Especially the KPP part may need some discussion.

## Suggested addition to `tag-index`
o introduce 3D field SWFrac3D that stores shortwave penetration rather than computing it multiple times
  - remove local solution in pkg/seaice
  - add CPP-option KPP_USE_SWFRAC3D, undefined by default, because this code can only approximately replace the pkg/kpp-code

jm-c commented:
@mjlosch Does this PR address,in part or fully, issue #263 ?

mjlosch commented:
>  Does this PR address,in part or fully, issue #263 ?

Yes, it does, and goes a little further (includes KPP). I was not aware of the issue, otherwise I would have linked it.

jm-c commented:
@mjlosch I think this implementation is likely to cause some confusion, at least for P-coordinate, since the 3D variable "SWFrac3D", which corresponds to a vertical interface,  is shifted by 1 level: SWFrac3D(k) corresponds to interface rF(k+1).

jm-c commented:
@mjlosch I am not sure that the size of "SWFrac3D" needs to be extended to "Nr+1" (but still think that there might be some confusion  in the way it's implemented when using P-coords). I suggest we chat a little bit about various options that we coud think of for this new array setting, and then decide,

menemenlis commented:
This is nice. We will probably need to use it at some point and combine it with the radiative transfer package in ECCO-Darwin. I just want to mention https://github.com/MITgcm-contrib/ecco_darwin/tree/master/regions/RedSea/v05_kpp --- a 2D implementation of swfrac (swfrac2d), which should eventually be superseded or combined with swfrac3d.

jm-c commented:
My recollection of what has been discussed (on Dec 05 and 06):

Regarding the size of "SWFrac3D": to be consistent with usual vertical indexing of fields located at grid-cell interface, both with Z-coord and P-coord, mentionned 2 solutions:
1. easiest is to extend the size to `Nr+1`. This way we can access the very top interface, at k=1 in Z-coord and at k=Nr+1 in P-coord, with consistent MITgcm k indexing.
2. the alternative would be to keep the same size (Nr) but only use the value from k=2 to Nr. This would requires to change the definition of  "SWFrac3D" to the penetrating short-wave fraction relative to surface penetrating short-wave, so that we can assume it's always one at the very top (and don't need to store it). Also the coding will be less simple that with first option (Nr+1) as we need to take care of the 2 special cases, at very top (= 1) and very bottom (=0).

It would be nice to add a new run-time switch to turn on/off the penetrating short-wave:

- make sense to keep the CPP option (with a new 3-D array coming from this PR);
- can choose an integer selector (instead of boolean) to allow to differentiate (in near future) between uniform + time independent (current code), spacialy varying + time independent, or interactive/prognostics (coupled with biology ...).
-  default could be ON when CPP option is defined (backward compatible).

@mjlosch If I forgot something or report incorrectly, please correct.

mjlosch commented:
@jm-c I went with option (1) and increased the vertical length of `SWFrac3D` to `Nr+1`. That makes a lot of things much easier, but requires the use `gravitySign` for the vertical divergence of the penetrating shortwave radiation.

For the runtime switch, things are not very clear to me. I can use a flag `usePenetratingSW`, that is already there (in comments) when ever `SWFrac3D` is used (this flag can also be `selectPenetratingSW` or similar and with different values 0,1,2, ...), but there are many instances of `SHORTWAVE_HEATING` code, that is just reading and applying `Qsw`. Using fortran-ifs in these case would make the code quite difficult, e.g. in `autodiff_store/restore.F`, `addummy_in_stepping.F`, `exf_getffields.F`, `exf_getforcing.F`, etc. I can give it a try, but I am sure that you will not agree with many of my solutions.

mjlosch commented:
`SHORTWAVE_HEATING` also appears in pkgs: autodiff, exf (a lot!), layers, long step, monitor, ocn_compon_interf.
Mostly, we probably don't have to do anything about those occurrences, but for exf, there may be situations where the runtime parameter `selectPenetratingSW` may be useful. I just don't know, if it makes sense, because often the CPP flag is combined with `ALLOW_ATM_TEMP` and the construction refers to `swflux` and not so much `Qsw`.

jm-c commented:
@mjlosch Thanks for taking the lead in adding a long awaited run-time switch there. 
I did not look at your recent changes but will check to see what does this means for other pkgs use of `SHORTWAVE_HEATING` to answer your question.

mjlosch commented:
Unfortunately the addition of the runtime switch changes the default OpenAD test `global_ocean.90x40x15` (looses some working precision).

jm-c reviewed (commented):
I did not look neither at `pkg/kpp` nor `pkg/seaice` but have few comments (mostly related to  the new parameter `selectPenetratingSW`):

1. I don't really like the changes in PARAMS.h that are not related to this PR. It was expected since it seems (git blame) that I wrote these lines the way I like.
2. I think we should keep loading of QswFile even if `selectPenetratingSW=0`,
    a. in case one specify `surfQFile` instead of `surfQnetFile`;
    b. some pkgs might use `Qsw `for something else than heating the ocean, e.g., pkg/dic 
 (option USE_QSW, parameter: parfrac, ...).
    This would mean that we don't need to add a condition on `selectPenetratingSW`  in
 few places (e.g., `external_fields_load.F`) and no changes to pkg/exf/.
3. we can also revert changes to `external_forcing.F` since it does compile without
  any modification and gives the same results (I tried in lab_sea.salt_plume).

I might have other comments once I finish looking at code changes and run few testreport.

jm-c commented on model/src/external_forcing_surf.F:
```diff
@@ -216,17 +216,27 @@ SUBROUTINE EXTERNAL_FORCING_SURF(
+C     it separately, we need to remove it from the temperature
+C     forcing. Because of the sign conventions, this means adding it
+C     here after it was substracted together with Qnet.
+        IF ( selectPenetratingSW .GE. 0 ) THEN
```
It looks like a mistake here. Should be `selectPenetratingSW .GE. 1`, I think.

mjlosch replied:
I wasn't sure, what to do here, but I guess you are right, if we consider that `selectPenetratingSW .EQ. 0` should be same as `#undef SHORTWAVE_HEATING`.

jm-c replied:
In `apply_forcing.F`, `Qsw` is added under the condition `selectPenetratingSW .GT. 0`. The same condition needs to be used here in `external_forcing_surf.F` to remove `Qsw`, otherwise the model does not receive the correct net heat flux `Qnet`.

mjlosch commented:
Thanks for the comments:

> 1. I don't really like the changes in PARAMS.h that are not related to this PR. It was expected since it seems (git blame) that I wrote these lines the way I like.

I was afraid of that (o:, if you insist, we can revert these changes, I was just trying to comply to the F77-80-character rule (in case we want to run the MITgcm on a punch-card machine). We usually do this and (the usual comment), often the default editor width is 80 char and you get these annoying line breaks.

> 2. I think we should keep loading of QswFile even if `selectPenetratingSW=0`,
>    a. in case one specify `surfQFile` instead of `surfQnetFile`;
>    b. some pkgs might use `Qsw `for something else than heating the ocean, e.g., pkg/dic
>    (option USE_QSW, parameter: parfrac, ...).
>    This would mean that we don't need to add a condition on `selectPenetratingSW`  in
>    few places (e.g., `external_fields_load.F`) and no changes to pkg/exf/.

Sounds good

> 3. we can also revert changes to `external_forcing.F` since it does compile without
>    any modification and gives the same results (I tried in lab_sea.salt_plume).

Is this, because `external_forcing.F` is deprecated?

jm-c commented:
Regarding `external_forcing.F`, the reason to keep this file was that, if a user has a customized version of it, this customized version could still be used (with the CPP option defined) and comparing the customized version with current version would help to move the changes into `apply_forcing.F`. So the fewer changes we add to `external_forcing.F` the better it is.

jm-c commented:
@mjlosch are you going to fix this part (the second point I raised on Mar 29):
> I think we should keep loading of QswFile even if selectPenetratingSW=0,
I have other minor comments regarding the filling of 3-D array `SWFrac3D` but it's more of a detail. I also see differences in testreport results that look tike machine truncation, except for TAF -AD `1D_ocean_ice_column` (but did not look at `pkg/seaice` changes).

mjlosch commented:
> @mjlosch are you going to fix this part (the second point I raised on Mar 29):
> 
> > I think we should keep loading of QswFile even if selectPenetratingSW=0,
> 
I can remove the condition `selectPenetratingSW.GE.0` from `external_fields_load.F` (revert the change), if that's what you mean (I added that in case we want to explicitly avoid loading `QswFile` by setting `selectPenetratingSW=-1` or so). As it is now, we do load `QswFile` also for the case of `selectPenetratingSW=0`.

> I have other minor comments regarding the filling of 3-D array `SWFrac3D` but it's more of a detail. 

Please go ahead and suggest (or even make) changes.

> I also see differences in testreport results that look tike machine truncation, except for TAF -AD `1D_ocean_ice_column` (but did not look at `pkg/seaice` changes).

In the forward experiments, I see changes in `global_ocean.cs32x15.icedyn/in_p/seaice/thsice`, all `lab_sea` experiments and all `seaice_obcs` experiments. Most of them are small, but I am not surprised that `pkg/seaice` is involved in all of them.

For `global_ocean.cs32x15.seaice` the changes are expected, because I fixed a bug (see above), for `global_ocean.cs32x15.in_p` the experiment fails now, too, but it was not very good to start with, still I should investigate:
```
< Y Y Y Y>10<16 16 16 16 13 16 16 16 13 13 12 14 13 13 13 14 22 14 13 16 22 14 13 16 pass  global_ocean.cs32x15.in_p
---
> Y Y Y Y> 8<11 13 10 10 12 16 12 10 11 11  7 10 11 10  8 10 22 11  6  7 22 11  7  7 FAIL  global_ocean.cs32x15.in_p
```

There's probably more to come in the AD-experiements.

mjlosch commented:
As I could have figured out from my own comments (instead I performed complicated tests), the changes in both `global_ocean.cs32x15.in_p` and `global_ocean.cs32x15.seaice` are a consequence of fixing the bug in `pkg/seaice`. The truncation errors stem from the change in `external_forcing_surf.F`, where I changed the order of operations to subtract `Qsw` from `Qnet` in the case of `selectPenetratingQsw .GT. 0`. We can change that back (revert `external_forcing_surf.F`) and always subtract `Qsw` if `SHORTWAVE_HEATING` is defined. I am tempted to do that, as the case of `selectPenetratingQsw .EQ. 0`, which I was trying to avoid, is probably only going to be used for tests, and hence is going to be rare in practice. @jm-c, what do you think?

jm-c commented:
@mjlosch I did not read carefully the condition on `selectPenetratingSW` in `external_fields_load.F`, sorry.
But allowing to use `selectPenetratingSW= -1` seems a little bit convoluted and at the riisk of beeing inconsistent with, e.g., `pkg/dic` and `surfQFile`, so might be cleaner to remove the condition on `selectPenetratingSW`.

And regarding removing avoiding the splitting of loops in `external_forcing_surf.F`, you could try, and if it does reduce output updates, then we could decide what to do.

mjlosch commented:
> @mjlosch I did not read carefully the condition on `selectPenetratingSW` in `external_fields_load.F`, sorry. But allowing to use `selectPenetratingSW= -1` seems a little bit convoluted and at the riisk of beeing inconsistent with, e.g., `pkg/dic` and `surfQFile`, so might be cleaner to remove the condition on `selectPenetratingSW`.

OK, I can do that.

> And regarding removing avoiding the splitting of loops in `external_forcing_surf.F`, you could try, and if it does reduce output updates, then we could decide what to do.

I tested this and this splitting of the loops is the only reason for the truncation level errors. I can check in that change for you to test (and we can revert it afterwards, if we don't like it). Did you mean that?

jm-c commented:
@mjlosch Yes, go for it. It could also help to figure out the output changes in TAF-AD tests.

mjlosch commented on model/src/diags_oceanic_surf_flux.F:
```diff
@@ -116,13 +116,19 @@ SUBROUTINE DIAGS_OCEANIC_SURF_FLUX( myTime, myIter, myThid )
      &      *HeatCapacity_Cp*rUnit2mass
           ENDDO
          ENDDO
+#ifdef SHORTWAVE_HEATING
+         IF ( selectPenetratingSW .GE. 0 ) THEN
+          DO j = 1,sNy
+           DO i = 1,sNx
+            tmp1k(i,j,bi,bj) = tmp1k(i,j,bi,bj) - Qsw(i,j,bi,bj)
+           ENDDO
+          ENDDO
+         ENDIF
+#endif
```
We could revert this file as well.

jm-c replied:
I am playing with `external_forcing_surf.F` a little bit, and depending on what we do there will affect what to do with this diagnostic.

jm-c replied:
I am going to push an update to `external_forcing_surf.F`, preserve the current output and does the right thing regarding `selectPenetratingSW `.

jm-c commented:
@mjlosch :
1. do you agree with my latest commit there in `external_forcing_surf.F` ? and if yes, do you want to adjust/fix `diags_oceanic_surf_flux.F` accordingly ?

2. The other thing that I would like to fix in current master branch SHORTWAVE_HEATING code is an inconsistency  (and heat loss or gain) when setting `seaLev_Z` (and therefore `rF(1)` ) to a non zero value. This feature is not much used (not tested) but would be nice to fix this problem in this (short-wave) PR. I could push something to `ini_forcing.F` if you like.

3. An other other thing regarding the setting of `SWFrac3D` with pCoords in `ini_forcing.F`: we could use a difference of `phiRef*recip_gravity` to get the depth (this eliminates the use of `rhoConst` in non-boussinesq setting). For now, I would like to add this line of code (using `phiRef`) but commented out, and can decide later.

mjlosch commented:
> 1. do you agree with my latest commit there in `external_forcing_surf.F` ? and if yes, do you want to adjust/fix `diags_oceanic_surf_flux.F` accordingly ?

I like this solution. I will carry this over to `diags_oceanic_surf_flux.F`.

> 2. The other thing that I would like to fix in current master branch SHORTWAVE_HEATING code is an inconsistency  (and heat loss or gain) when setting `seaLev_Z` (and therefore `rF(1)` ) to a non zero value. This feature is not much used (not tested) but would be nice to fix this problem in this (short-wave) PR. I could push something to `ini_forcing.F` if you like.

Fixing problems is always a good idea, so go for it.

> 3. An other other thing regarding the setting of `SWFrac3D` with pCoords in `ini_forcing.F`: we could use a difference of `phiRef*recip_gravity` to get the depth (this eliminates the use of `rhoConst` in non-boussinesq setting). For now, I would like to add this line of code (using `phiRef`) but commented out, and can decide later.

Here you are referring to line 157 `SWFracK(k) = -rF(k)*recip_rhoConst*recip_gravity`, aren't you? Please push your stuff, if it does not change results, I am don't see any problem adding this now.

jm-c commented:
@mjlosch my third point above was not exactly right (since rhoConst is used to set phiRef), but I still added this line of code, commented out, as an alternative. 
And, as expected, the changes in `ini_forcing.F` are not affecting any results (good).
Also, I did not realized earlier that the `KPP_OPTIONS.h` was added to exp. 1D_ocean_ice_column `code_ad` dir, but now the significant changes that I see for the AD test make sense.

jm-c commented:
Regarding next things to do (some we talked about yesterday), from my side:

1. in `pkg/kpp`, a linear interpolation would be better than a level-based step function.
2. in `kpp_check.F`, should remove the STOP when CPP option SHORTWAVE_HEATING is undef (see issue #913).
3. In `config_summary.F`, I will move the report of `selectPenetratingSW` closer to other forcing parameters.
4. could discuss later which test experiment to use to test KPP_USE_SWFRAC3D
5. I would be OK to keep `cfc_example` unchanged, meaning that the updated code in `layer_thermodynamics.F` will not be compiled (same as before for the short-wave heating code there).

jm-c commented:
I am going to make changes (in this branch) in `pkg/kpp` so that it can be used with `#undef SHORTWAVE_HEATING`. After that could re-visit the changes specific to 3-D Qsw.

Two things I noticed in `pkg/kpp`: does not compile with `#define KPP_USE_SWFRAC3D` and `#undef SHORTWAVE_HEATING`. And the vertical averaging in `kpp_subroutined.F`, line 478:
> worka(i) = 0.5*( swatt(i,kl) + swatt(i,kl-1) )

looks suspicious.

mjlosch commented:
> I am going to make changes (in this branch) in `pkg/kpp` so that it can be used with `#undef SHORTWAVE_HEATING`. After that could re-visit the changes specific to 3-D Qsw.
> 
> Two things I noticed in `pkg/kpp`: does not compile with `#define KPP_USE_SWFRAC3D` and `#undef SHORTWAVE_HEATING`. And the vertical averaging in `kpp_subroutined.F`, line 478:
> 
> > worka(i) = 0.5*( swatt(i,kl) + swatt(i,kl-1) )
> 
> looks suspicious.

Great, I'll wait for your changes. The compile problem should be easy to sort out, especially once (or if) we've turned `KPP_USE_SWFRAC3D` into a runtime parameter. For the interpolation of `swatt`, I assumed that `zgrid=rC` (see `kpp_init_fixed.F`), but it should be `worka(i) = 0.5*( swatt(i,kl) + swatt(i,kl+1) )`, maybe a higher order interpolation would be in place, as we know that the profile is mostly exponential.

mjlosch commented:
About the interpolation of `swFrac3D` to cell centres (in l478 in `kpp_routines.F`, but also when interpolating to `hbl` in `kpp_calc.F`, ll632): The exponential function will always be concave, and  I want to used only already computed values of swFrac3D. I tried Taylor expansions of the type `f(rF(k)-delta) = f(k) - delta f(k)'`, with `drF(k)>delta>0`, which should always underestimate the true value (because of the exponential shape of `f`). The best I can do (for some reason, that I don't understand), to average `f(rF(k)-delta) = f(k) - delta f(k)'` and `f(rF(k+1)+drC(k)-delta) = f(k+1) + (drC(k)-delta) f(k+1)'` and then assume that the derivatives are the same central differences `f(k)'=f(k+1)'=(f(k)-f(k+1))/drF(k)` to arrive at 
```
0.5*(f(k)+f(k+1)) + (0.5 - delta/drF(k))*(f(k)-f(k+1))
```
This very well for `k>1`, for `k=1` the errors are large for anything I do, but as far as I can see, we don't really have that situation. In `kpp_routines.F`, `k>1` (why?) in `kpp_calc.F`, `KPPfrac` is computed from `hbl`, which can be as small as `0.5*drF(1)` in the surface layer, and then it is only used once for the counter gradient transport with `ghat`. I am not sure how important errors/inaccuracies are in this case.

jm-c commented:
I made changes to allow to use `pkg/kpp` with "#undef SHORTWAVE_HEATING" (and selectPenetratingSW=0 ):
1. I changed what is stored in `KPPfrac` to be 1 minus previous `KPPfrac` so that it is now the fraction of `Qsw` heating the mixing layer (instead of going through the mixing layer). This simplifies a little bit where it's used (in `kpp_transport_t.F`) and works better for both cases with and without short-wave heating (no short-wave, no heating, keep it at zero).
2. I put the calls to S/R SWFRAC within "#ifdef SHORTWAVE_HEATING", it's not strickly neccessary but this will be needed for `SWFrac3D`, so it makes sense to also have these SWFRAC calls within the same #ifdef
3. I did not really test this, just run Fwd and Adm testreport for all tests that have #define SHORTWAVE_HEATING", don't see any changes. Also run the 2 kpp `vermix` test with default `CPP_OPTIONS.h` (i.e., "#undef SHORTWAVE_HEATING").

mjlosch commented:
@jm-c, your changes look good to me.

I tried to replace the new CPP-flag `KPP_USE_SWFRAC3D` by a run-time parameter and ran into the following problem: If we want to keep CPP-flag `SHORTWAVE_HEATING` (because it involves defining an external 3D field `SWfrac3D`), we can only pass `SWfrac3D` to `s/r kpp_mix` if `SHORTWAVE_HEATING` is defined (otherwise the 3D field is not defined). I see two options:
1. replace `KPP_USE_SWFRAC3D` by `SHORTWAVE_HEATING` in the parameter list of `s/r kpp_mix` and leave the rest of the code as it is. 
2. do not pass `SWfrac3D` ever to `s/r kpp_mix`. Instead, replace the relevant parts of the `SHORTWAVE_HEAT` code by a new `s/r kpp_swfrac(kl,z,swatt,...)` or  similar, where `kl` is the current level (or -1, in which case one would need to figure out, at which level we are), `z` is the actual depth (e.g. `zgrid(kl)`, `hbl`) and do the linear interpolation according to `swatt = f(k) - delta/drF(k)*(f(k)-f(k+1))` within that routine and return the appropriate `swatt`. This routine would include `GRID.h`, `FFIELDS.h`, etc., and it could use a new runtime flag `KPPuseSWfrac3D` locally to decide whether we want to use `s/r swfrac` or somehow interpolate from `SWFRAC3D`.

In my opinion, option 2 would be more elegant, but might change default results due to truncation errors. What do you think?

jm-c commented:
@mjlosch I prefer option (1), which is more in line with how `pkg/kpp` is written, also avoid many model common block in higher level S/R (currently in `kpp_subroutines.F`).  This is also the how the `pkg/salt_plume` interface in KPP is written.

mjlosch commented:
OK, I'll look into that. It will mean, however, that the interpolation code will have to be duplicated 2 or 3 times, as far as I can see.

jm-c commented:
@mjlosch If you feel it's better to go with solution (2) to avoid repeating the interpolation code, then you should go with this solution. We can always revisit the need to include header files vs passing few variables as arguments. Plus I might have missed something along the way, so it was not a strong preference.

mjlosch commented:
@jm-c I lost track, are you done with your change, or do I still need to wait?

jm-c commented:
@mjlosch I am done with making changes for now. Thanks.

mjlosch commented:
@jm-c KPP_USE_SWFRAC3D is history and hence AD `1D_ocean_ice_column` has gone back to default. We can still decide to use the new runtime flag `KPPuseSWfrac3D` in that experiment, but I am more interested in the AD-code generation, which now takes place in `1D_ocean_ice_column`, `lab_sea`, and `tutorial_tracer_adjsens` (because no cpp-flag needs to be defined). This also caused quite few problems mostly with the re-use of `worka` in the `SHORTWAVE_HEATING` code and the `ALLOW_SALT_PLUME` code. This got fixed with `$taf incomplete worka` statements and some store statements, but I guess it's pretty fragile (w.t.r. AD-code generation) and needs a watchful eye in the future. Maybe we can change the interface of `swfrac` in some future PR to not override the input (depth) by the totally different output (fraction) to make this code more robust in AD generation. 

For the interpolation, I went with option (1), i.e. no new subroutine, also because of difficulties with `swfrac` and `worka`. We might want to increase the order of interpolation for better accuracy. For the interpolation, I assumed that `zgrid` is always at the center of the cell (which not always true). If we don't want to do that, we need to introduce a field that contains `rF`.

I think that this code is now ready for review.

jm-c commented on pkg/kpp/kpp_calc.F:
```diff
@@ -551,13 +550,19 @@ SUBROUTINE KPP_CALC(
 #endif /* SALT_PLUME_SPLIT_BASIN */
 #endif /* ALLOW_SALT_PLUME */
-     I       dbloc, Ritop, work2,
+     I       dbloc, Ritop, fCori(1-OLx,1-OLy,bi,bj),
```
`work2` was "_RL" array but `fCori` is "_RS" so this will not work unless S/R KPPMIX is changed (which does not seems to be the case right now).

There is also a similar issue with passing "_RS"  array `SWFRac3D` to S/R KPPMIX as this later one is expecting a "_RL" variable.

jm-c commented on pkg/kpp/kpp_calc.F:
```diff
@@ -195,10 +196,15 @@ SUBROUTINE KPP_CALC(
       _RL vddiff( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy, 0:Nrp1, mdiff )
       _RL ghat  ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy, Nr            )
       _RL hbl   ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy                )
+      INTEGER kbl(1-OLx:sNx+OLx, 1-OLy:sNy+OLy                )
```
I have the impression that this variable "kbl" is not used when SHORTWAVE_HEATING is "#undef". Did I missed something ?

jm-c commented on pkg/kpp/kpp_calc.F:
```diff
@@ -618,35 +623,62 @@ SUBROUTINE KPP_CALC(
+#ifdef SHORTWAVE_HEATING
+      IF ( selectPenetratingSW .GE. 1 ) THEN
+#ifdef ALLOW_AUTODIFF_TAMC
+C     Avoid ueslessly recomputing this block because of re-using worka
```
a typo in "ueslessly" ?

jm-c commented on pkg/kpp/kpp_calc.F:
```diff
@@ -618,35 +623,62 @@ SUBROUTINE KPP_CALC(
+        DO j=1-OLy,sNy+OLy
+         DO i=1-OLx,sNx+OLx
+          k = kbl(i,j)
+          rFac = MAX( (KPPhbl(i,j,bi,bj)+rF(k))/drF(k), zeroRL )
```
Should should use "*recip_drF" instead of "/drF" ?

mjlosch commented on pkg/kpp/kpp_calc.F:
```diff
@@ -618,35 +623,62 @@ SUBROUTINE KPP_CALC(
+          k = kbl(i,j)
+          rFac = MAX( (KPPhbl(i,j,bi,bj)+rF(k))/drF(k), zeroRL )
+          worka(i,j) = SWFrac3D(i,j,k,bi,bj)
+     &         - rFac*(SWFrac3D(i,j,k,bi,bj) - SWFrac3D(i,j,k+1,bi,bj))
```
The same is true in two cases in `kpp_routines.F`

mjlosch commented on pkg/kpp/kpp_calc.F:
```diff
@@ -200,14 +190,26 @@ SUBROUTINE KPP_CALC(
       _RL TTALPHA( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy, Nrp1 )
       _RL SSBETA ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy, Nrp1 )
 cph)
+#if ( defined ALLOW_SALT_PLUME || defined SHORTWAVE_HEATING )
+      _RL worka ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy                )
+#endif
+#ifdef ALLOW_SALT_PLUME
+      _RL temp1   ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy, Nr          )
+      _RL temp2   ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy, Nr          )
+      _RL boplume ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy, Nrp1        )
+#ifdef SALT_PLUME_SPLIT_BASIN
+      _RL lon ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy                  )
+      _RL lat ( 1-OLx:sNx+OLx, 1-OLy:sNy+OLy                  )
+#endif /* SALT_PLUME_SPLIT_BASIN */
+#endif /* ALLOW_SALT_PLUME */
```
should we fix the alignment of `(`, `)`? also with previous lines?

jm-c replied:
Could leave it as it is, does not matter too much.

mjlosch commented:
I don't have anything to add. Let's merge!

jm-c commented:
OK, I am going to merge this PR soon. This will also close issue #913 and issue #263 (will add a comment there).

jm-c commented:
I found few little details in my "notes" that I forgot about, and could push here (last minute ?) if @mjlosch you agree:

1. in `ini_forcing.F`, rename "kp" to "km" (since it's for the mask)
2. in `config_summary.F` move the report of `selectPenetratingSW` just after `addFrictionHeating` (easier to find in STDOUT).
3. I am in favor of not adding the new `CPP_OPTIONS.h` in `cfc_example/code` (I don't mind if `layer_thermodynamics.F` is compiled without  SHORTWAVE_HEATING defined there).

mjlosch commented:
@jm-c I am fine with all of your "last-minute" suggestions. Go ahead and push them.
